### PR TITLE
tcp_proxy: better handling flow control given the upstream can be multiplexed

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:
Tcp_proxy can use upstream h2/h3 as upstream. Since h2 and h3 streams can be multiplexed in the upstream connection,
there are more than 1 sources of upstream write throttling control. One from upstream connection stream buffer and one from
upstream h2/h3 stream buffer.

Additional Description:

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by lambdai** [comment]
I didn't find the existing flow control test cases for h2/h3 upstream tunnel. 
@alyssawilk Could you advice where the best place to add?

**Comment by lambdai** [comment]
> I didn't find the existing flow control test cases for h2/h3 upstream tunnel. @alyssawilk Could you advice where the best place to add?

The reduced test coverage is partial because some corner cases are not covered in test cases yet. Is that something we can ignore in this PR? It's just some log

File: source/common/tcp_proxy/tcp_proxy.cc [hunk 1]
```diff
@@ -274,22 +274,48 @@ void Filter::UpstreamCallbacks::onEvent(Network::ConnectionEvent event) {
...
-    parent_->readDisableDownstream(true);
+  if (above_high_watermark_counter_++ == 0) {
+    if (parent_ != nullptr) {
+      ENVOY_CONN_LOG(
+          trace,
+          "tcp_proxy disables downstream read because upstream write buffer is now above high "
+          "watermark. above_high_watermark_counter_ = {}",
+          *parent_->downstreamConnection(), above_high_watermark_counter_);
+      // There's too much data buffered in the upstream write buffer, so stop reading.
+      parent_->readDisableDownstream(true);
```
Comment (alyssawilk) [inline source/common/tcp_proxy/tcp_proxy.cc:285]: the stats in read disable downstream assume calling disable pauses or resumes reading.
I'd suggest either only calling readDisabledownstream when this transitions, or updating readDisableDownstream to check internally.

**Review by alyssawilk — COMMENTED** [review]
Ah good catch.  I'd add large buffer backup tests to the tcp tunneling integration tests.
/wait

**Pull request closed without merging**